src/vec/FixedPairList.cpp

```diff
@@ -0,0 +1,345 @@
+
+  LOG4ESPP_LOGGER(FixedPairList::theLogger, "FixedPairList");
+
+  FixedPairList::FixedPairList(shared_ptr< espressopp::storage::Storage > storage)
+    : globalPairs()
+  {
+    LOG4ESPP_INFO(theLogger, "construct FixedPairList");
+
+    if(!storage->getSystem()->vectorization) {
+      throw std::runtime_error("system has no vectorization");
+    }
+    vectorization = storage->getSystem()->vectorization;
+
+    if(!(vectorization->storageVec))
+      throw std::runtime_error("vectorization->storageVec cannot be null");
+    auto& storageVec = vectorization->storageVec;
+
+    sigBeforeSend = storage->beforeSendParticles.connect(
+      boost::bind(&FixedPairList::beforeSendParticles, this, _1, _2));
+    sigAfterRecv = storage->afterRecvParticles.connect(
+      boost::bind(&FixedPairList::afterRecvParticles, this, _1, _2));
+    sigOnParticlesChanged = storage->onParticlesChanged.connect(
+      boost::bind(&FixedPairList::onParticlesChanged, this));
+
+    if(vectorization->getVecLevel() != 2) {
+      throw std::runtime_error("espressopp::vec::FixedPairList can only be used for vecLevel=2");
+    }
+  }
+
+  FixedPairList::~FixedPairList()
+  {
+    LOG4ESPP_INFO(theLogger, "~FixedPairList");
+    sigBeforeSend.disconnect();
+    sigAfterRecv.disconnect();
+    sigOnParticlesChanged.disconnect();
+  }
+
+  real FixedPairList::getLongtimeMaxBondSqr()
+  {
+    return longtimeMaxBondSqr;
+  }
+
+  void FixedPairList::setLongtimeMaxBondSqr(real d)
+  {
+    longtimeMaxBondSqr = d;
+  }
+
+  void FixedPairList::resetLongtimeMaxBondSqr()
+  {
+    longtimeMaxBondSqr = 0.0;
+  }
+
+  bool FixedPairList::add(size_t pid1, size_t pid2)
+  {
+    bool returnVal = true;
+    if (pid1 > pid2)
+      std::swap(pid1, pid2);
+
+    auto const& storageVec = vectorization->storageVec;
+    size_t const p1 = storageVec->lookupRealParticleVec(pid1);
+    size_t const p2 = storageVec->lookupLocalParticleVec(pid2);
+
+    if(p1==VEC_PARTICLE_NOT_FOUND) {
+      // Particle does not exist here, return false
+      returnVal=false;
+    } else {
+      if(p2==VEC_PARTICLE_NOT_FOUND) {
+        LOG4ESPP_DEBUG(theLogger, "Particle p2 " << pid2 << " not found");
+      }
+    }
+
+    if(returnVal){
+      // add the pair locally
+      this->push_back({p1,p2});
+
+      // ADD THE GLOBAL PAIR
+      // see whether the particle already has pairs
+      std::pair<GlobalPairs::const_iterator,
+        GlobalPairs::const_iterator> equalRange
+        = globalPairs.equal_range(pid1);
+      if (equalRange.first == globalPairs.end()) {
+        // if it hasn't, insert the new pair
+        globalPairs.insert(std::make_pair(pid1, pid2));
+      }
+      else {
+        // otherwise test whether the pair already exists
+        for (GlobalPairs::const_iterator it = equalRange.first; it != equalRange.second; ++it) {
+          if (it->second == pid2) {
+            // Pair already exists, generate error!
+            throw std::runtime_error("Duplicate pair inserted");
```

Comment:
Have you included this check also in the other versions of FixedPairList?

Comment is unnecessary at this point.

Reply:
> Have you included this check also in the other versions of FixedPairList?
No, still a TODO there.

> Comment is unnecessary at this point.
Edited.